### ISSUE TITLE
Add CMake to test suite

### ISSUE THREAD

**Issue by dmed256** [libocca/occa#325]
## Description

We just switched to using Github Actions for our CI and it would be nice to add a new step that uses CMake to build the library and tests. The file can be found here:
- https://github.com/libocca/occa/blob/master/.github/workflows/build.yml

**Comment by SFrijters** [comment]
I would like to help with this, but I have no experience with GitHub Actions yet.

**Comment by dmed256** [comment]
@SFrijters Sweet!

The 2 "steps" used to compile the library + tests are:

```yml
    - name: Compile library
      run: make -j

    - name: Compile tests
      run: make -j tests
```

We would need to change it to something like

```yml
    - name: Compile library
      run: make -j
      if: ${{ !matrix.useCMake }}

    - name: Compile tests
      run: make -j tests
      if: ${{ !matrix.useCMake }}

    - name: Compile library (CMake)
      run: ???
      if: ${{ matrix.useCMake }}

    - name: Compile tests (CMake)
      run: ???
      if: ${{ matrix.useCMake }}
```

The new `useCMake` variable would come from the `matrix` definition

```yaml
      matrix:
        include:
          - name: (Ubuntu) gcc-8
            os: ubuntu-18.04
            CC: gcc-8
            CXX: g++-8
            OCCA_COVERAGE: 1
            useCMake: false             # <--- New variable useCMake defined here

          - name: (Ubuntu) CMake + gcc-8
            os: ubuntu-18.04
            CC: gcc-8
            CXX: g++-8
            OCCA_COVERAGE: 1
            useCMake: true              # <--- New variable useCMake defined here
```

**Comment by gionellef** [comment]
Hi, I'm a beginner contributor who's willing to learn from this by doing. I've looked at the issue, and... since there are CMakeLists.txt in place already and the task is supposed to just call it, is it enough to just ?

`run: cmake ./`

Is there a need to generate the cmakefiles in a separate subdirectory "build"? 

I hope I understood the requirement correctly :)

**Comment by dmed256** [comment]
@gionellef Hey, thanks for reaching out and wanting to help out!

Right now there is no CMake test in the [.github/workflows/build.yml](https://github.com/libocca/occa/blob/master/.github/workflows/build.yml) file. It would be nice to have one Github Actions job that would do instead of

```yml
    - name: Compiler info
      run: make -j info

    - name: Compile library
      run: make -j

    - name: Compile tests
      run: make -j tests

    - name: Run unit tests
      run: ./tests/run_tests

    - name: Run examples
      run: ./tests/run_examples
```

something like

```yml
    - name: Compile library
      run: <cmake command>

    - name: Compile tests
      run: <cmake command>

    - name: Run unit tests
      run: ./tests/run_tests

    - name: Run examples
      run: ./tests/run_examples
```

It might be a really simple task now that @SFrijters added `CMakeLists.txt` throughout the codebase. But I'm not too familiar with `cmake` so I was hoping to get help with this task :)

> Is there a need to generate the cmakefiles in a separate subdirectory "build"?
I think so? The few times I've used `cmake` I've built the project in a separate directory (e.g. `./build`) but up to you.

**Comment by SFrijters** [comment]
Ah yes, sorry, this fell off my radar for a bit. From memory (not at work at the moment), I do something like

```
mkdir build   # Make build dir so we don't pollute the source tree outside this one directory
cd build
cmake ../     # Additional CMake options go here like -DALLOW_CUDA:BOOL=ON -DCMAKE_INSTALL_PREFIX:PATH=/install/dir/here -DENABLE_TESTS:BOOL=ON
make -j       # Build using multiple cores
ctest         # Run the test suite
make install  # Install
```

**Comment by gionellef** [comment]
Alright, so I tried the simplest way that works in my local machine, which is what @SFrijters just mentioned. cmake ../ make-j and so on, however, the pipeline isn't having any of it. Compilation error [undefined reference to `omp_get_num_threads'](https://github.com/libocca/occa/pull/342/checks?check_run_id=699039560)

I believe it's a missing target_link_library somewhere in the CMakeLists, I'm not sure, but I don't understand github actions well enough to know if there are any other differences between running it on my machine versus on a github runner. :(

**Comment by SFrijters** [comment]
Interesting... can you try with `-DALLOW_OPENMP:BOOL=ON` passed to CMake? I think I've used the code exclusively with OpenMP enabled so far, so maybe there's some reference that's not properly guarded by `#ifdef`s.

**Comment by SFrijters** [comment]
I didn't have as much time this week as I'd hoped - is this still an issue? I don't see this particular error show up in the latest build log for your PR?

**Comment by gionellef** [comment]
You're right, it looks like compilation is successful already but CTest isn't, with one test failing.

```
55/90 Test #55: test-io-utils ..............................Child aborted***Exception:   0.36 sec
      Start 56: test-lang-expr
```

I removed the ctest part in my latest commit which caused the run examples step to now fail. However, without the entire `Compile tests (CMake)` step, the build finishes successfully.

**Comment by SFrijters** [comment]
Could you add the `--output-on-failure` flag to the cmake call? That should let it tell us what is wrong. It's probably good to keep that flag around, not just for digging into this particular failure.
Depending on what comes out I can take a look at that next week hopefully (no promises this time); I recall that test-io-utils was failing for me as well at some point (there is a `#ifndef USE_CMAKE` in there now).

**Comment by dmed256** [comment]
Fixed in #425
